File: F-OMP/cfd.f90
program cfd

  use boundary
  use jacobi
  use cfdio

  use omp_lib

  implicit none

! Output frequency
  
  integer, parameter :: printfreq = 1000

! Variables associated with convergence

  double precision :: error, bnorm

! Set tolerance for convergence; zero or negative means do not check

  double precision, parameter :: tolerance = 0.0d0

! Main arrays

  double precision, allocatable ::  psi(:,:), zet(:,:)
  double precision, allocatable ::  psitmp(:,:), zettmp(:,:)

! Command-line arguments

  integer :: scalefactor,  numiter

  double precision :: re  ! re = 3.7 seems to be stability limit with Jacobi

  integer, parameter :: maxline = 32
  character(len=maxline) :: tmparg

!  Basic sizes of simulation

  integer, parameter :: bbase = 10
  integer, parameter :: hbase = 15
  integer, parameter :: wbase =  5
  integer, parameter :: mbase = 32
  integer, parameter :: nbase = 32

  logical :: irrotational = .true., checkerr = .false.

!  Some auxiliary parameters and variables

  integer :: m, n, b, h, w
  integer :: iter

  integer :: nthread

  double precision :: tstart, tstop, ttot, titer, modvsq, hue

!  Are we stopping based on tolerance?

  if (tolerance .gt. 0.0) checkerr = .true.

!  Parallel initialisation

!  Read in parameters

  if (command_argument_count() /= 2 .and. command_argument_count() /= 3) then

     write(*,*) 'Usage: cfd <scale> <numiter> [reynolds]'
     stop

  end if

  call get_command_argument(1, tmparg)
  read(tmparg,*) scalefactor
  call get_command_argument(2, tmparg)
  read(tmparg,*) numiter

  if (command_argument_count() == 3) then

     irrotational = .false.
     call get_command_argument(3, tmparg)
     read(tmparg,*) re
        
  else

     re = -1.0
     
  end if

  if (.not. checkerr) then
     write(*,fmt='('' Scale factor = '',i3,'', iterations = '', i6)') &
           scalefactor, numiter
  else
     write(*,fmt='('' Scale factor = '',i3,'', iterations = '', i6, &
          &'', tolerance = '', g11.4)') scalefactor, numiter, tolerance
  end if

  if (irrotational) then
        
     write(*,*) 'Irrotational flow'
        
  else

     write(*,fmt='('' Reynolds number = '', f6.3)') re
        
  end if

!  Calculate b, h & w and m & n
        
  b = bbase*scalefactor 
  h = hbase*scalefactor
  w = wbase*scalefactor 
  m = mbase*scalefactor
  n = nbase*scalefactor

  re = re / dble(scalefactor)

  nthread = omp_get_max_threads()

  write(*,fmt='('' Running CFD on '', i4, '' x '', i4, '' grid using '', &
                  &i4, '' thread(s)'')') m, n, nthread

!  Allocate arrays, including halos on psi and tmp

  allocate(psi(0:m+1, 0:n+1))
  allocate(zet(0:m+1, 0:n+1))

  allocate(psitmp(0:m+1, 0:n+1))

  if (.not. irrotational) then

     allocate(zettmp(0:m+1, 0:n+1))

  end if

!  Zero the psi array

  psi(:,:) = 0.0
  zet(:,:) = 0.0

!  Set the psi boundary condtions which are constant

   call boundarypsi(psi, m, n, b, h, w)

!  Compute normalisation factor for error

   bnorm = sum(psi(:,:)**2)

   if (.not. irrotational) then

!    Update the zeta boundary condtions which depend on psi

     call boundaryzet(zet, psi, m, n)

!    Update the normalisation

     bnorm = bnorm + sum(zet(:,:)**2)

  end if

   bnorm = sqrt(bnorm)

!  Begin iterative Jacobi loop

   write(*,*)
   write(*,*) 'Starting main loop ...'
   write(*,*)

   tstart = gettime()

  do iter = 1, numiter

!  Compute the new psi based on the old one

     if (irrotational) then

!  Call function with no vorticity

        call jacobistep(psitmp, psi, m, n)

     else

!  Call function containing vorticity

        call jacobistepvort(zettmp, psitmp, zet, psi, m, n, re)

     end if

!  Compute current error value if required
     
     if (checkerr .or. iter == numiter) then

        error = deltasq(psitmp, psi, m, n)

        if (.not. irrotational) then

           error = error + deltasq(zettmp, zet, m, n)

        end if

        error = sqrt(error)
        
        error = error / bnorm

     end if

!  Copy back

!$omp parallel workshare
     psi(1:m, 1:n) = psitmp(1:m, 1:n)
!$omp end parallel workshare

     if (.not. irrotational) then

!$omp parallel workshare
        zet(1:m, 1:n) = zettmp(1:m, 1:n)
!$omp end parallel workshare

     end if

     if (.not. irrotational) then

!    Update the zeta boundary condtions which depend on psi

        call boundaryzet(zet, psi, m, n)
        
     end if

!  Quit early if we have reached required tolerance

     if (checkerr) then
        if (error .lt. tolerance) then
           write(*,*) 'CONVERGED iteration ', iter, ': terminating'
           exit
        end if
     end if

!  End iterative Jacobi loop

     if (mod(iter,printfreq) == 0) then

        if (.not. checkerr) then
           write(*,*) 'completed iteration ', iter
        else
           write(*,*) 'completed iteration ', iter, ', error = ', error
        end if

     end if

  end do

  if (iter .gt. numiter) iter = numiter

  tstop = gettime()

  ttot  = tstop-tstart
  titer = ttot/dble(iter)

  write(*,*) 
  write(*,*) '... finished'
  write(*,*)
  write(*,fmt='('' After    '', i6, '' iterations, error is '', g11.4)') &
        iter, error
  write(*,fmt='('' Time for '', i6, '' iterations was '',&
        &g11.4, '' seconds'')') iter, ttot
  write(*,fmt='('' Each individual iteration took '', g11.4, '' seconds'')') &
        titer
  write(*,*)
  write(*,*) 'Writing output file ...'

!  Output results

  call writedatafiles(psi, m, n, scalefactor)

!  Output gnuplot file

  call writeplotfile(m, n, scalefactor)

! Finish

  write(*,*) ' ... finished'
  write(*,*)
  write(*,*) 'CFD completed'
  write(*,*)

end program cfd

File: C-OMP/cfd.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

#include <omp.h>

#include "arraymalloc.h"
#include "boundary.h"
#include "jacobi.h"
#include "cfdio.h"

int main(int argc, char **argv)
{
  int printfreq=1000; //output frequency
  double error, bnorm;
  double tolerance=0.0; //tolerance for convergence. <=0 means do not check

  //main arrays
  double **psi, **zet;
  //temporary versions of main arrays
  double **psitmp, **zettmp;

  //command line arguments
  int scalefactor, numiter;

  double re; // Reynold's number - must be less than 3.7

  //simulation sizes
  int bbase=10;
  int hbase=15;
  int wbase=5;
  int mbase=32;
  int nbase=32;

  int irrotational = 1, checkerr = 0;

  int m,n,b,h,w;
  int iter;
  int i,j;

  // OpenMP variables

  int nthread;

  double tstart, tstop, ttot, titer;

  //do we stop because of tolerance?
  if (tolerance > 0) {checkerr=1;}

  //check command line parameters and parse them

  if (argc <3|| argc >4)
    {
      printf("Usage: cfd <scale> <numiter> [reynolds]\n");
      return 0;
    }

  scalefactor=atoi(argv[1]);
  numiter=atoi(argv[2]);

  if (argc == 4)
    {
      re=atof(argv[3]);
      irrotational=0;
    }
  else
    {
      re=-1.0;
    }

  if(!checkerr)
    {
      printf("Scale Factor = %i, iterations = %i\n",scalefactor, numiter);
    }
  else
    {
      printf("Scale Factor = %i, iterations = %i, tolerance= %g\n",scalefactor,numiter,tolerance);
    }

  if (irrotational)
    {
      printf("Irrotational flow\n");
    }
  else
    {
      printf("Reynolds number = %f\n",re);
    }

  //Calculate b, h & w and m & n
  b = bbase*scalefactor;
  h = hbase*scalefactor;
  w = wbase*scalefactor;
  m = mbase*scalefactor;
  n = nbase*scalefactor;

  re = re / (double)scalefactor;

  nthread = omp_get_max_threads();
  
  printf("Running CFD on %d x %d grid using %d thread(s)\n",m,n,nthread);

  //allocate arrays

  psi    = (double **) arraymalloc2d(m+2,n+2,sizeof(double));
  psitmp = (double **) arraymalloc2d(m+2,n+2,sizeof(double));

  //zero the psi array
  for (i=0;i<m+2;i++)
    {
      for(j=0;j<n+2;j++)
	{
	  psi[i][j]=0.0;
	}
    }

  if (!irrotational)
    {
      //allocate arrays

      zet =   (double **) arraymalloc2d(m+2,n+2,sizeof(double));
      zettmp =(double **) arraymalloc2d(m+2,n+2,sizeof(double));

      //zero the zeta array

      for (i=0;i<m+2;i++)
	{
	  for(j=0;j<n+2;j++)
	    {
	      zet[i][j]=0.0;
	    }
	}
    }
  
  //set the psi boundary conditions

  boundarypsi(psi,m,n,b,h,w);

  //compute normalisation factor for error

  bnorm=0.0;

  for (i=0;i<m+2;i++)
    {
      for (j=0;j<n+2;j++)
	{
	  bnorm += psi[i][j]*psi[i][j];
	}
    }

  if (!irrotational)
    {
      //update zeta BCs that depend on psi
      boundaryzet(zet,psi,m,n);

      //update normalisation

      for (i=0;i<m+2;i++)
	{
	  for (j=0;j<n+2;j++)
	    {
	      bnorm += zet[i][j]*zet[i][j];
	    }
	}
    }

  bnorm=sqrt(bnorm);

  //begin iterative Jacobi loop

  printf("\nStarting main loop...\n\n");
  
  tstart=gettime();

  for(iter=1;iter<=numiter;iter++)
    {
      //calculate psi for next iteration

      if (irrotational)
	{
	  jacobistep(psitmp,psi,m,n);
	}
      else
	{
	  jacobistepvort(zettmp,psitmp,zet,psi,m,n,re);
	}

      //calculate current error if required

      if (checkerr || iter == numiter)
	{
	  error = deltasq(psitmp,psi,m,n);

	  if(!irrotational)
	    {
	      error += deltasq(zettmp,zet,m,n);
	    }

	  error=sqrt(error);
	  error=error/bnorm;
	}

      //copy back

#pragma omp parallel for default(none) private(i,j) shared(psi,psitmp,m,n)
      for(i=1;i<=m;i++)
	{
	  for(j=1;j<=n;j++)
	    {
	      psi[i][j]=psitmp[i][j];
	    }
	}

      if (!irrotational)
	{
#pragma omp parallel for default(none) private(i,j) shared(zet,zettmp,m,n)
	  for(i=1;i<=m;i++)
	    {
	      for(j=1;j<=n;j++)
		{
		  zet[i][j]=zettmp[i][j];
		}
	    }
	}

      if (!irrotational)
	{
	  //update zeta BCs that depend on psi
	  boundaryzet(zet,psi,m,n);
	}

      //quit early if we have reached required tolerance

      if (checkerr)
	{
	  if (error < tolerance)
	    {
	      printf("Converged on iteration %d\n",iter);
	      break;
	    }
	}

      //print loop information

      if(iter%printfreq == 0)
	{
	  if (!checkerr)
	    {
	      printf("Completed iteration %d\n",iter);
	    }
	  else
	    {
	      printf("Completed iteration %d, error = %g\n",iter,error);
	    }
	}
    }

  if (iter > numiter) iter=numiter;

  tstop=gettime();

  ttot=tstop-tstart;
  titer=ttot/(double)iter;


  //print out some stats

  printf("\n... finished\n");
  printf("After %d iterations, the error is %g\n",iter,error);
  printf("Time for %d iterations was %g seconds\n",iter,ttot);
  printf("Each iteration took %g seconds\n",titer);

  //output results

  writedatafiles(psi,m,n, scalefactor);

  writeplotfile(m,n,scalefactor);

  //free un-needed arrays
  free(psi);
  free(psitmp);

  if (!irrotational)
    {
      free(zet);
      free(zettmp);
    }

  printf("... finished\n");

  return 0;
}
